### DISCUSSION TITLE
Circular Dependency TypeORM - Up to date doc examples?

### DISCUSSION THREAD

**Discussion by chronos1nside** [nartc/mapper#355] [Q&A]
First of all, thank you for this awesome project, it's been a life savior and reducer of code haha

So, here's the thing...

I'm running into some issues when mapping from and to TypeORM entities, mainly with the relations between two or more entities and I have a question:

**Is this section of the documentation up to date?** https://automapperts.netlify.app/docs/plugins-system/classes-limitations/

I'm only asking because I'm trying to follow the approach of "mapWith()" to try to avoid circular dependency using TypeOrm Entities but I keep facing some errors, even with the examples provided.

```
mapper.createMap(SomeOrderDto, Order).forMember(
  (d) => d.person,
  mapWith(
    () => Person, ****
    (s) => s.person,
    () => SomePersonDto
  )
);
```
This is what I got:
**** **Argument of type '() => typeof Person' is not assignable to parameter of type 'string | Constructor<unknown>'**

I tried something like this, but "lost" the mapper in the process.

```
mapper.createMap(SomeOrderDto, Order).forMember(
        (d) => d.person,
        mapWith(Person, SomeOrderDto, (s) => s.person),
      );
```

I'm running the latests version of all the packages and on a NestJS project, also updated, so I'm good with that.

I hope I made myself clear if not, feel free to ask questions.

Thanks in advance.

**Comment by nartc** [comment]
Hi @chronos1nside , thank you for the kind words. This is a tricky situation (circular deps is always tricky) so would you like to setup a call so we can brainstorm? 

And yes, that portion in the docs is outdated :(. I'll make an update asap

**Reply by chronos1nside** [reply]
Hi and I'm sorry for the late reply (busy with work) :(

Unfortunately, I can't get on a call as of now, but I manage to workaround my issue; here is what I did:

Keep in mind that I'm using the entities that are on the documentation as an example.

If I follow the "normal" approach, which it is that I keep "@AutoMap()" on all attributes of my classes (entities) that I wanted and needed to be mapped, I'm welcomed with the following message when I call the ".save()" of typeORM.

"Internal Server Error: undefined, Cannot perform update query because update values are not defined. Call "qb.set(...)" method to specify updated values."

The .save in this case, is just an insert but it is misunderstood as an update by typeORM.

As soon as I remove the "AutoMap" from the relation attributes (OneToMany - Order Entity Class) my insert works just fine, but the updates queries that use .save() are now broken.

So, to solve my issue, I created two profiles, one for creation only and the other for general updates and find queries; all attributes on the entity classes kept their "@AutoMap" annotation.

```
mapper.updateMap(SomePersonDto, Person)
mapper.updateMap(SomeOrderDto, Order)
```

```
mapper.createMap(SomePersonDto, Person);
mapper.createMap(SomeOrderDto, Order).forMember(
     (d) => d.person,
     mapFrom((s) => undefined),
);
```
As you can see, since I kept all my "AutoMap" annotations, I just removed the ones that I didn't need on creation cases.

TypeORM kept on leaving some primary and foreign keys missing, so I just had to call the missing relations directly to the database (small  inconvenience, but not a breaking one).

```
order.person = await this.personEntity.findOne({
     where: { person },
});
await this.orderEntity.save(order);
```

I hope that I kept things clear.

**Reply by nartc** [reply]
Thank you for the explanation. I'm still not convinced that `@AutoMap` would interfere with TypeORM, but then again, I'm not using TypeORM as much. That's why I'd love to jump in a call so we can figure it out. I'm glad you found a workaround but if you have time, let's schedule a call. If you don't have time, that's fine. Thanks again!
